modules/optimization/include/executioners/OptimizeSolve.h

```diff
@@ -0,0 +1,135 @@
+{
+public:
+  static InputParameters validParams();
+  OptimizeSolve(Executioner & ex);
+
+  virtual bool solve() override;
+
+  const OptimizationReporter & getOptimizationReporter() const { return *_obj_function; }
+
+  /**
+   * Record tao data for output by a reporter
+   */
+  void getTaoSolutionStatus(std::vector<int> & tot_iters,
+                            std::vector<double> & gnorm,
+                            std::vector<int> & obj_iters,
+                            std::vector<double> & cnorm,
+                            std::vector<int> & grad_iters,
+                            std::vector<double> & xdiff,
+                            std::vector<int> & hess_iters,
+                            std::vector<double> & f,
+                            std::vector<int> & tot_solves) const;
+
+protected:
+  /// Bounds routine
+  virtual PetscErrorCode variableBounds(Tao tao);
+
+  /// Objective routine
+  virtual Real objectiveFunction();
+
+  /// Gradient routine
+  virtual void gradientFunction(libMesh::PetscVector<Number> & gradient);
+
+  /// Hessian application routine
+  virtual PetscErrorCode applyHessian(libMesh::PetscVector<Number> & s,
+                                      libMesh::PetscVector<Number> & Hs);
+
+  /// Communicator used for operations
+  const libMesh::Parallel::Communicator _my_comm;
+
+  /// List of execute flags for when to solve the system
+  const ExecFlagEnum & _solve_on;
+
+  /// objective function defining objective, gradient, and hessian
+  OptimizationReporter * _obj_function = nullptr;
+
+  /// Tao optimization object
+  Tao _tao;
+
+private:
+  bool _verbose;
+  /// tao solver info
+  int _obj_iterate = 0;
+  int _grad_iterate = 0;
+  int _hess_iterate = 0;
+  std::vector<int> _total_iterate_vec;
+  std::vector<int> _obj_iterate_vec;
+  std::vector<int> _grad_iterate_vec;
+  std::vector<int> _hess_iterate_vec;
+  std::vector<int> _function_solve_vec;
+  std::vector<double> _f_vec;
+  std::vector<double> _gnorm_vec;
+  std::vector<double> _cnorm_vec;
+  std::vector<double> _xdiff_vec;
+
+  /// Here is where we call tao and solve
+  PetscErrorCode taoSolve();
+
+  /// output optimization iteration solve data
+  static PetscErrorCode monitor(Tao tao, void * ctx);
+
+  void setTaoSolutionStatus(double f, int its, double gnorm, double cnorm, double xdiff);
+
+  ///@{
+  /// Function wrappers for tao
+  static PetscErrorCode objectiveFunctionWrapper(Tao tao, Vec x, Real * objective, void * ctx);
+  static PetscErrorCode hessianFunctionWrapper(Tao tao, Vec x, Mat hessian, Mat pc, void * ctx);
+  static PetscErrorCode applyHessianWrapper(Mat H, Vec s, Vec Hs);
+  static PetscErrorCode
+  objectiveAndGradientFunctionWrapper(Tao tao, Vec x, Real * objective, Vec gradient, void * ctx);
+  static PetscErrorCode variableBoundsWrapper(Tao /*tao*/, Vec xl, Vec xu, void * ctx);
+  ///@}
+
+  // fixme lynn these have weird names because Petsc already has macros for TAONTR etc.
```

Comment:
remove fixme

if names add weird translate them to what TAO people think they are

Reply:
done